utils/awsFunctions.ts

```diff
@@ -0,0 +1,136 @@
+   * @param {string} email - The email to associate with the user.
+   * @returns {Promise<any>} - The AWS response or error.
+   */
+  export const createUser = async (username: string, email: string): Promise<any> => {
+    const client = configureAWSCredentials();
+  
+    const params: CreateUserCommandInput = {
+      IdentityStoreId: identityStoreId,
+      UserName: username,
+      Name: {
+        Formatted: username,
+        FamilyName: username,
+        GivenName: username,
+      },
+      DisplayName: username,
+      Emails: [
+        {
+          Value: email,
+          Type: "work",
+          Primary: true,
+        },
+      ],
+    };
+  
+    try {
+      const command = new CreateUserCommand(params);
+      return (await client.send(command));
+    } catch (error) {
+      console.error(`The error from create user ${error}`);
+      throw new Error(`Failed to create user: ${error instanceof Error ? error.message : String(error)}`);
+    }
+  };
+  
+  /**
+   * This function adds the user to the IAM group.
+   * @param {string} groupId - The group ID to add the user to.
+   * @param {string} userId - The user ID to add to the group.
+   * @returns {Promise<any>} - The AWS response or error.
+   */
+export const addUserToGroup = async (groupId: string, awsUserId: string): Promise<any> => {
+    const client = configureAWSCredentials();
+  
+    const params: CreateGroupMembershipCommandInput = {
+      IdentityStoreId: identityStoreId,
+      GroupId: groupId,
+      MemberId: {
+        UserId: awsUserId,
+      },
+    };
+  
+    try {
+      const command = new CreateGroupMembershipCommand(params);
+      return (await client.send(command));
+    } catch (error) {
+      console.error("Error adding user to group:", error);
+      if (error.__type === 'ConflictException'){
+        return { conflict: true };
+      }
+
+      throw new Error(`Failed to add user to group: ${error instanceof Error ? error.message : String(error)}`);
+    }
+  };
+  
+  /**
+   * Function to get UserId by username from AWS Identity Store.
+   * @param {string} username - The username of the user.
+   * @returns {Promise<string | null>} - The UserId if found, otherwise null.
+   */
+  export const fetchAwsUserIdByUsername = async (username: string): Promise<string | null> => {
+    const client = configureAWSCredentials();
+  
+    const params: ListUsersCommandInput = {
+      IdentityStoreId: identityStoreId,
+      Filters: [
+        {
+          AttributePath: "UserName",
+          AttributeValue: username,
+        },
+      ],
+    };
+    
+    let response;
+    try {
+      response = await client.send(new ListUsersCommand(params));
+    } catch (err) {
+      throw new Error(`Error while fetching user by username: ${err instanceof Error ? err.message : String(err)}`);
+    }
+    if (response.Users && response.Users.length > 0) {
+      const userId = response.Users[0].UserId!;
```

Comment:
The user will be at 0 index?

Reply:
yes, since the command is list command, it will get the list of the users, we are making sure the users are having unique name at the time of creation

Reply:
Because the chances of user name of a user in RDS backend changing is less, so we can assume that the username will be in sync at both places, AWS and RDS backend will have same usernames

Reply:
> It returns user's list. But, how are we sure that we will return userId from user which is at 0 index against the username passed as argument to the function?

I didn't find any direct function to fetch the user it was only possible to fetch via filters in list command

Reply:
<img width="897" alt="Screenshot 2024-11-11 at 5 19 09 PM" src="https://github.com/user-attachments/assets/52cd578a-9309-46d6-b42d-93248a07b472">
Correct me if I am wrong, the response contains the users list but it will have only user as it fetches with filter you mentioned.

Reply:
Yes correct